Deprecate calling `redirect_to` with bad status

tjefferson08 opened the pull request:
### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

TL;DR `redirect_to` and `assert_redirect_to` both currently accept non-redirect status codes

### Detail

Currently, `redirect_to` and `assert_redirected_to` will gladly accept "nonsensical" status values, e.g. `:bad_request`. I say nonsensical because using `redirect_to` implies that you intend to, well, redirect! But browsers (and turbo) will not process these responses as redirects because their HTTP response status falls outside the 3XX range

### Additional information
This PR adds a deprecation warning to `redirect_to` calls with non-redirect statuses. It does so by asking the response if it is a `redirect?` immediately after assigning the response status.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included

zzak commented:
> A potentially much better option would be to change ActionController::Redirecting#redirect_to to raise an error upon encountering a non-3XX.

So the option you took was to raise an exception in the test if the redirect_to uses a non-standard status, but that is assuming there is even test coverage. I'm pretty sure what we want (as documented) is to deprecate accepting a non-3XX status and then raise in a future version of Rails. Would you be willing to try that?

tjefferson08 commented:
>deprecate accepting a non-3XX status and then raise in a future version of Rails

makes sense to me 👍 I'll take a crack at that

tjefferson08 commented:
Ok I've taken a pass at it! Decided to try and leverage the `redirect?` helper on `ActionDispatch::Response`; I was trying to avoid duplicating the status code list provided by `Rack::Response::Helpers`

cc @zzak

zzak commented on actionpack/lib/action_controller/metal/redirecting.rb:
```diff
@@ -107,6 +107,7 @@ def redirect_to(options = {}, response_options = {})
       allow_other_host = response_options.delete(:allow_other_host) { _allow_other_host }
 
       self.status = _extract_redirect_to_status(options, response_options)
+      ActionController.deprecator.warn("Passing `redirect_to` a status other than a valid 3XX redirect is deprecated and will be removed in a future version of Rails") unless redirect?
```
```suggestion
    unless redirect?
      ActionController.deprecator.warn <<~MSG
        Passing a non-3XX status to `redirect_to` is deprecated and will be removed in Rails 8.1.
      MSG
    end
```

Indentation looks off, but something like this, maybe.

zzak replied:
I think we want to raise in the future, not remove / no-op.

tjefferson08 replied:
of course! thanks, I got a little mixed up looking at other examples of `deprecator.warn`

thanks for the support @zzak! I've updated the warning message 🚀

tjefferson08 commented on actionpack/lib/action_controller/metal/redirecting.rb:
```diff
@@ -105,9 +105,15 @@ def redirect_to(options = {}, response_options = {})
 
+      unless redirect?
+        ActionController.deprecator.warn <<~MSG
+          Passing a non-3XX status to `redirect_to` is deprecated and will raise
+          an exception in Rails 8.1.
```
it does split across two lines. sounds like we want one line then? I've added a squish

seanpdoyle commented on actionpack/lib/action_controller/metal/redirecting.rb:
```diff
@@ -105,9 +105,15 @@ def redirect_to(options = {}, response_options = {})
-
       self.status = _extract_redirect_to_status(options, response_options)
 
+      unless redirect?
```
Is it possible to achieve this behavior without delegating the `#redirect?` predicate?

[ActionController::Metal](https://edgeapi.rubyonrails.org/classes/ActionController/Metal.html) defines a `#response` attribute. Is that the same instance as `@_response`?

```suggestion
      unless response.redirect?
```

tjefferson08 replied:
thanks for the input!

I originally started with what you've suggested; it works 👍 

Hm it probably does make sense to keep from widening `ActionController::Metal` surface area; upon closer inspection there's nothing else in the API that looks "convenience-y" like `redirect?`

happy to make that change

tjefferson08 committed:
Deprecate calling `redirect_to` with bad status

`redirect_to` should only be called with valid redirect status codes

It was convenient here to defer the question of "which status codes are
valid for redirects?" to `ActionDispatch::Response#redirect?` via
`Rack::Response::Helpers`
